### ISSUE TITLE
Need to continuously test published Python packages in isolation

### ISSUE THREAD

**Issue by feuGeneA** [0xProject/0x-monorepo#2066]
## Expected Behavior

We should have a continuously updated indicator of whether each of our Python packages tests cleanly when installed fresh from PyPI.

## Current Behavior

Currently our CI tests exercise the copies of packages that are checked in to git, not the copies that are published to PyPI.

Also, we have tests of published packages that can be run manually (command-line in local development environment), but they are not integrated with any automated process.

## Possible Solution

??? Please help! :grinning: 

## Context

After the release of `0x-contract-wrappers`, I was doing some extended testing, and I found problems.  Our CI testing did not catch these problems because they do not test published packages.

We do have a good _way_ to run tests against deployed packages (using `tox`, with config files that are already in place) but we don't have a good _time_ to run them.  They can be run manually in a local development environment, but it doesn't make much sense to run them in our monorepo CI runs, since what needs to be tested against is in PyPI, not the source repository.

To clarify, I'll explain two ways we have of running `tox`.  (There may be more ways, but these are what we're currently set up for.)

The first is to use `tox` in it's normal way, which is for it to run against a single package (of which there are many in our monorepo), and for it to download from PyPI all of that package's dependencies and install them in a clean "virtual environment", then install (the local copy of) the package-under-test into that virtual environment, and then finally to run the given commands (eg tests, but could be anything) in that environment.  This doesn't work in our monorepo setup because often times changes to package P, checked in to branch B, also depend on changes to package Q, which are also checked into that same branch B.  So, testing package P using this `tox` approach would fail because it would use the version of package Q that's published to PyPI, not the version that's in branch B.

The other (currently configured) way to run `tox` is almost the same as above, except that instead of using the local copy of the package-under-test, it uses the copy that's published to PyPI.  So, this one makes even _less_ sense to run in our monorepo CI runs, since it uses _no_ code in the git-branch-under-test.

So the big question is: where *should* we be automating these tests?  Should we maybe set something up that's automatically triggered by a publish to PyPI instead of a commit to Git?  That seems reasonable, but are there existing CI solutions that can handle this type of setup?  Does CircleCI do anything like that?  Does _any_ CI service do anything like that?  Or would we have to spin up our own Jenkins instance or something like that?  What other approaches could we consider taking?

Depending on manual, local-dev-env runs of these tests will inevitably result in errors in published packages...

**Comment by fabioberger** [comment]
Thanks for opening this issue @feuGeneA. For our Typescript packages, we ran into the same challenge. We ran into issues with our packages that only became apparent after publishing to NPM (e.g., missing type declarations, improper scoping of `npmignore`'d packages, etc...).

What we were able to do was setup `test-publish`, a CI test that publishes the current code in the repo to a local Verdaccio repository (i.e., self-hosted NPM registry), and then install them from there in a clean-slate environment. Is there an Verdaccio equivalent for PyPI?

**Comment by feuGeneA** [comment]
I'll have to look for a python analog to Verdaccio, but we may be able to leverage test.pypi.org. It's a hosted service just like the real pypi (same upload tools work with it, etc), but it's just for testing. It (and the real pypi) won't let you upload the same version twice, so we'd have to permute the version "number" with every run, maybe by appending a git commit cash to the version number or something like that. It could potentially get pretty convoluted, but it's an option.

**Comment by opaolini** [comment]
If I understand your issue correctly, you aim to test the published packages and those tests are unrelated to what is checked into the `0x-monorepo`. So ideal workflow would be:
- The `0x-monorepo` Ci finishes, and publishes a new version of a package.
- A webhook fires (probably PyPI does not have any webhooks per-se) and triggers a CI pipeline via an API

As I mentioned to @fabioberger out of all the CICD tools out there, I am the least familiar with Circle, however as is the case with [GitlabCI](https://docs.gitlab.com/ee/ci/triggers/) CircleCI seems to [also have triggers](https://circleci.com/docs/2.0/triggers/) - why not leverage that? 

Basically we'd have an external service, that monitors the packages on PyPI (or any package repository in the future), and if it detects a new version it triggers a build on the CI.

However, I've noticed that `CircleCI` does not allow for trigger variables that could be passed in (for example which version to test), so that would require the test itself to handle that.

This could potentially answer the "where" and "when" :)

**Comment by feuGeneA** [comment]
Thanks @opaolini . I love the idea of a service monitoring the package repo and triggering builds upon new publications.  What might the implementation and hosting of such a service look like?

And, where might we have a good information radiator for the status of such builds?  Builds associated with a pull request are nicely prominent on the PR page, which has SOMEone (author, reviewer, etc) looking at it throughout the PR's lifecycle.  But there's no analogous, often-viewed page associated with a particular package. We would have a status badge, of course, on the PyPI page, package-level README.md, etc, but I'm afraid that could go unnoticed.  Maybe just an email alert? Or Slack integration? Or all of the above? Any other ideas?

**Comment by opaolini** [comment]
I think it would be best to implement such a scraper on our own, as it doesn't seem like too big of an endeavour, and I could gladly come up with something. as It would be also a good field test of our "future self-hosted CI solution" 😄 

As for the CI status, I think all of those places you have mentioned are valid. We could have a slack bot which sends a message when a new package has been published, and also reports its CI status once that finishes. And for the builds that fail we could send an email / direct slack message to the maintainers.

You've mentioned that you run those tests manually, if you could maybe give me an example of that for one of the packages, I could put all of the information in a task for me 🙂
